File: gateway_unified/src/claude_gateway/stream.py
import json
from typing import Any, Dict, List, Tuple


def infer_synthetic_block_from_event(event_type: str, event_data: Dict[str, Any]) -> Dict[str, Any]:
    """根据 SSE 事件推断缺失的 content_block_start 应该使用的 block 类型。"""
    if event_type == "content_block_delta":
        delta = event_data.get("delta")
        if isinstance(delta, dict):
            delta_type = delta.get("type")
            if delta_type == "input_json_delta":
                return {"type": "tool_use", "id": "", "name": "", "input": {}}
        return {"type": "text", "text": ""}
    if event_type == "content_block_stop":
        return {"type": "text", "text": ""}
    return {"type": "text", "text": ""}


def normalize_input_json_for_stream(raw: str) -> Tuple[str, str]:
    """将分片的 tool input JSON 合并为完整 JSON。"""
    text = raw if isinstance(raw, str) else ""
    if not text.strip():
        return "{}", "empty"

    try:
        parsed = json.loads(text)
        return json.dumps(parsed, ensure_ascii=False, separators=(",", ":")), "parsed_full"
    except Exception:
        pass

    decoder = json.JSONDecoder()
    pos = 0
    fragments: List[Any] = []

    while pos < len(text):
        while pos < len(text) and text[pos].isspace():
            pos += 1
        if pos >= len(text):
            break
        try:
            obj, end = decoder.raw_decode(text, pos)
        except Exception:
            fragments = []
            break
        fragments.append(obj)
        pos = end

    if fragments:
        merged: Any = fragments[0]
        for obj in fragments[1:]:
            if isinstance(merged, dict) and isinstance(obj, dict):
                merged.update(obj)
            elif isinstance(merged, list) and isinstance(obj, list):
                merged.extend(obj)
            else:
                merged = obj
        return json.dumps(merged, ensure_ascii=False, separators=(",", ":")), f"merged_fragments:{len(fragments)}"

    return json.dumps({"raw": text}, ensure_ascii=False, separators=(",", ":")), "fallback_wrapped_raw"


def emit_frame(event_name: str | None, data_payload: str | None) -> bytes:
    """构造一个 SSE 帧。"""
    out_lines: List[str] = []
    if event_name:
        out_lines.append(f"event: {event_name}")
    if data_payload is not None:
        for segment in data_payload.split("\n"):
            out_lines.append(f"data: {segment}")
    return ("\n".join(out_lines) + "\n\n").encode("utf-8")


async def process_sse_frame(
    frame_lines: List[str],
    seen_block_starts: set,
    block_kind_by_index: Dict[int, str],
    tool_partial_json_by_index: Dict[int, str],
    pending_event_name: str | None,
) -> Tuple[List[bytes], str | None]:
    """
    处理一个 SSE 帧，返回 (输出字节列表, 更新后的 pending_event_name)。

    处理逻辑：
    - 补全缺失的 content_block_start（synthetic shim）
    - 缓存 tool_use 的 input_json_delta，在 content_block_stop 时合并发出
    - 处理跨帧的 event/data 分离
    """
    output_chunks: List[bytes] = []
    chunk_count = 0

    if not frame_lines:
        return output_chunks, pending_event_name

    event_name: str | None = None
    data_lines: List[str] = []
    for frame_line in frame_lines:
        if frame_line.startswith("event:"):
            if event_name is None:
                event_name = frame_line[len("event:"):].strip()
        elif frame_line.startswith("data:"):
            data_value = frame_line[len("data:"):]
            if data_value.startswith(" "):
                data_value = data_value[1:]
            data_lines.append(data_value)

    # 处理跨帧的 event/data 分离
    if event_name and not data_lines:
        return output_chunks, event_name
    if not event_name and pending_event_name and data_lines:
        event_name = pending_event_name
        pending_event_name = None
    if event_name and pending_event_name and data_lines:
        pending_event_name = None

    if not data_lines:
        output_chunks.append(("\n".join(frame_lines) + "\n\n").encode("utf-8"))
        return output_chunks, pending_event_name

    payload = "\n".join(data_lines)
    if payload == "[DONE]":
        output_chunks.append(emit_frame(event_name, payload))
        return output_chunks, pending_event_name

    try:
        parsed = json.loads(payload)
    except Exception as parse_exc:
        print(f"[gateway malformed sse data] {type(parse_exc).__name__}: {parse_exc}; payload_len={len(payload)}")
        safe_error = json.dumps(
            {
                "type": "error",
                "error": {
                    "type": "gateway_bad_event",
                    "message": "Dropped malformed upstream data event",
                },
            },
            ensure_ascii=False,
        )
        output_chunks.append(emit_frame("error", safe_error))
        return output_chunks, pending_event_name

    event_type = parsed.get("type") if isinstance(parsed, dict) else None
    event_index = parsed.get("index") if isinstance(parsed, dict) else None

    # 补全缺失的 content_block_start
    if (
        event_type in {"content_block_delta", "content_block_stop"}
        and isinstance(event_index, int)
        and event_index not in seen_block_starts
    ):
        synthetic_block = infer_synthetic_block_from_event(event_type, parsed)
        synthetic_payload = {
            "type": "content_block_start",
            "index": event_index,
            "content_block": synthetic_block,
        }
        print(
            "[gateway sse shim] synthetic_start=1 "
            f"trigger_event={event_type} index={event_index} "
            f"synthetic_type={synthetic_block.get('type')}"
        )
        output_chunks.append(emit_frame("content_block_start", json.dumps(synthetic_payload, ensure_ascii=False)))
        seen_block_starts.add(event_index)
        block_kind_by_index[event_index] = str(synthetic_block.get("type", "text"))
    elif event_type == "content_block_start" and isinstance(event_index, int):
        seen_block_starts.add(event_index)
        if isinstance(parsed.get("content_block"), dict):
            cb_type = parsed["content_block"].get("type")
            if isinstance(cb_type, str):
                block_kind_by_index[event_index] = cb_type

    # 缓存 tool_use 的 input_json_delta，延迟到 content_block_stop 时合并
    if (
        event_type == "content_block_delta"
        and isinstance(event_index, int)
        and block_kind_by_index.get(event_index) == "tool_use"
    ):
        delta = parsed.get("delta")
        if isinstance(delta, dict) and delta.get("type") == "input_json_delta":
            partial_json = delta.get("partial_json")
            if isinstance(partial_json, str):
                tool_partial_json_by_index[event_index] = (
                    tool_partial_json_by_index.get(event_index, "") + partial_json
                )
                print(
                    "[gateway sse shim] buffered_tool_input_delta=1 "
                    f"index={event_index} chunk_len={len(partial_json)} "
                    f"total_len={len(tool_partial_json_by_index[event_index])}"
                )
                return output_chunks, pending_event_name

    # content_block_stop 时发出合并后的 tool input JSON
    if (
        event_type == "content_block_stop"
        and isinstance(event_index, int)
        and block_kind_by_index.get(event_index) == "tool_use"
        and event_index in tool_partial_json_by_index
    ):
        normalized_json, normalize_reason = normalize_input_json_for_stream(
            tool_partial_json_by_index[event_index]
        )
        shim_delta_payload = {
            "type": "content_block_delta",
            "index": event_index,
            "delta": {
                "type": "input_json_delta",
                "partial_json": normalized_json,
            },
        }
        print(
            "[gateway sse shim] emitted_normalized_tool_input_delta=1 "
            f"index={event_index} reason={normalize_reason} normalized_len={len(normalized_json)}"
        )
        output_chunks.append(emit_frame("content_block_delta", json.dumps(shim_delta_payload, ensure_ascii=False)))
        tool_partial_json_by_index.pop(event_index, None)

    output_chunks.append(emit_frame(event_name, payload))
    return output_chunks, pending_event_name

File: gateway_unified/claude-gateway-rs/src/stream.rs
#![allow(dead_code)]

use serde_json::{Map, Value};
use std::collections::{HashMap, HashSet};

pub fn infer_synthetic_block_from_event(event_type: &str, event_data: &Value) -> Value {
    if event_type == "content_block_delta" {
        if let Some(Value::Object(delta)) = event_data.get("delta") {
            if let Some(Value::String(delta_type)) = delta.get("type") {
                if delta_type == "input_json_delta" {
                    let mut out = Map::new();
                    out.insert("type".to_string(), Value::String("tool_use".to_string()));
                    out.insert("id".to_string(), Value::String("".to_string()));
                    out.insert("name".to_string(), Value::String("".to_string()));
                    out.insert("input".to_string(), Value::Object(Map::new()));
                    return Value::Object(out);
                }
            }
        }
        let mut out = Map::new();
        out.insert("type".to_string(), Value::String("text".to_string()));
        out.insert("text".to_string(), Value::String("".to_string()));
        return Value::Object(out);
    }
    if event_type == "content_block_stop" {
        let mut out = Map::new();
        out.insert("type".to_string(), Value::String("text".to_string()));
        out.insert("text".to_string(), Value::String("".to_string()));
        return Value::Object(out);
    }

    let mut out = Map::new();
    out.insert("type".to_string(), Value::String("text".to_string()));
    out.insert("text".to_string(), Value::String("".to_string()));
    Value::Object(out)
}

pub fn normalize_input_json_for_stream(raw: &str) -> (String, String) {
    if raw.trim().is_empty() {
        return ("{}".to_string(), "empty".to_string());
    }

    if let Ok(parsed) = serde_json::from_str::<Value>(raw) {
        return (
            serde_json::to_string(&parsed).unwrap_or_default(),
            "parsed_full".to_string(),
        );
    }

    let stream = serde_json::Deserializer::from_str(raw).into_iter::<Value>();
    let mut fragments = Vec::new();
    for item in stream {
        if let Ok(val) = item {
            fragments.push(val);
        } else {
            fragments.clear();
            break;
        }
    }

    if !fragments.is_empty() {
        let mut merged = fragments.remove(0);
        for obj in fragments.iter() {
            if let (Value::Object(m), Value::Object(o)) = (&mut merged, obj.clone()) {
                for (k, v) in o {
                    m.insert(k, v);
                }
            } else if let (Value::Array(m), Value::Array(o)) = (&mut merged, obj.clone()) {
                m.extend(o);
            } else {
                merged = obj.clone();
            }
        }
        return (
            serde_json::to_string(&merged).unwrap_or_default(),
            format!("merged_fragments:{}", fragments.len() + 1),
        );
    }

    let mut fallback = Map::new();
    fallback.insert("raw".to_string(), Value::String(raw.to_string()));
    (
        serde_json::to_string(&Value::Object(fallback)).unwrap_or_default(),
        "fallback_wrapped_raw".to_string(),
    )
}

pub fn emit_frame(event_name: Option<&str>, data_payload: Option<&str>) -> Vec<u8> {
    let mut out_lines = Vec::new();
    if let Some(en) = event_name {
        out_lines.push(format!("event: {}", en));
    }
    if let Some(dp) = data_payload {
        for segment in dp.split('\n') {
            out_lines.push(format!("data: {}", segment));
        }
    }

    let mut res = out_lines.join("\n");
    res.push_str("\n\n");
    res.into_bytes()
}

pub struct SseState {
    pub seen_block_starts: HashSet<i64>,
    pub block_kind_by_index: HashMap<i64, String>,
    pub tool_partial_json_by_index: HashMap<i64, String>,
    pub pending_event_name: Option<String>,
}

impl SseState {
    pub fn new() -> Self {
        Self {
            seen_block_starts: HashSet::new(),
            block_kind_by_index: HashMap::new(),
            tool_partial_json_by_index: HashMap::new(),
            pending_event_name: None,
        }
    }
}

pub fn process_sse_frame(
    frame_lines: &[String],
    state: &mut SseState,
) -> (Option<String>, Vec<String>) {
    let mut output_data = Vec::new();
    if frame_lines.is_empty() {
        return (None, output_data);
    }

    let mut event_name: Option<String> = None;
    let mut data_lines = Vec::new();

    for line in frame_lines {
        if let Some(rest) = line.strip_prefix("event:") {
            if event_name.is_none() {
                event_name = Some(rest.trim().to_string());
            }
        } else if let Some(rest) = line.strip_prefix("data:") {
            let data_value = rest.strip_prefix(' ').unwrap_or(rest);
            data_lines.push(data_value.to_string());
        }
    }

    if event_name.is_some() && data_lines.is_empty() {
        state.pending_event_name = event_name.clone();
        return (None, output_data);
    }

    if event_name.is_none() && state.pending_event_name.is_some() && !data_lines.is_empty() {
        event_name = state.pending_event_name.clone();
        state.pending_event_name = None;
    }

    if event_name.is_some() && state.pending_event_name.is_some() && !data_lines.is_empty() {
        state.pending_event_name = None;
    }

    if data_lines.is_empty() {
        // 纯文本帧（如注释行），直接透传
        output_data.push(frame_lines.join("\n"));
        return (event_name, output_data);
    }

    let payload = data_lines.join("\n");
    if payload == "[DONE]" {
        output_data.push(payload);
        return (event_name, output_data);
    }

    let parsed_res: Result<Value, _> = serde_json::from_str(&payload);
    let parsed = match parsed_res {
        Ok(v) => v,
        Err(e) => {
            println!(
                "[gateway malformed sse data] Error: {}; payload_len={}",
                e,
                payload.len()
            );
            let mut error_obj = Map::new();
            error_obj.insert("type".to_string(), Value::String("error".to_string()));
            let mut err_inner = Map::new();
            err_inner.insert(
                "type".to_string(),
                Value::String("gateway_bad_event".to_string()),
            );
            err_inner.insert(
                "message".to_string(),
                Value::String("Dropped malformed upstream data event".to_string()),
            );
            error_obj.insert("error".to_string(), Value::Object(err_inner));
            return (
                Some("error".to_string()),
                vec![serde_json::to_string(&error_obj).unwrap()],
            );
        }
    };

    let event_type = parsed.get("type").and_then(|v| v.as_str());
    let event_index = parsed.get("index").and_then(|v| v.as_i64());

    if let (Some(ev_type), Some(ev_index)) = (event_type, event_index) {
        if (ev_type == "content_block_delta" || ev_type == "content_block_stop")
            && !state.seen_block_starts.contains(&ev_index)
        {
            let synthetic_block = infer_synthetic_block_from_event(ev_type, &parsed);
            let mut synthetic_payload = Map::new();
            synthetic_payload.insert(
                "type".to_string(),
                Value::String("content_block_start".to_string()),
            );
            synthetic_payload.insert(
                "index".to_string(),
                Value::Number(serde_json::Number::from(ev_index)),
            );

            let block_type = synthetic_block
                .get("type")
                .and_then(|v| v.as_str())
                .unwrap_or("text")
                .to_string();
            synthetic_payload.insert("content_block".to_string(), synthetic_block);

            println!(
                "[gateway sse shim] synthetic_start=1 trigger_event={} index={} synthetic_type={}",
                ev_type, ev_index, block_type
            );

            output_data.push(serde_json::to_string(&synthetic_payload).unwrap());
            state.seen_block_starts.insert(ev_index);
            state.block_kind_by_index.insert(ev_index, block_type);
        } else if ev_type == "content_block_start" {
            state.seen_block_starts.insert(ev_index);
            if let Some(Value::Object(cb)) = parsed.get("content_block") {
                if let Some(Value::String(cb_type)) = cb.get("type") {
                    state.block_kind_by_index.insert(ev_index, cb_type.clone());
                }
            }
        }
    }

    if let (Some(ev_type), Some(ev_index)) = (event_type, event_index) {
        if ev_type == "content_block_delta" {
            if let Some(kind) = state.block_kind_by_index.get(&ev_index) {
                if kind == "tool_use" {
                    if let Some(Value::Object(delta)) = parsed.get("delta") {
                        if delta.get("type").and_then(|v| v.as_str()) == Some("input_json_delta") {
                            if let Some(Value::String(partial_json)) = delta.get("partial_json") {
                                let entry = state
                                    .tool_partial_json_by_index
                                    .entry(ev_index)
                                    .or_default();
                                entry.push_str(partial_json);

                                println!("[gateway sse shim] buffered_tool_input_delta=1 index={} chunk_len={} total_len={}",
                                    ev_index, partial_json.len(), entry.len());
                                return (event_name, output_data);
                            }
                        }
                    }
                }
            }
        }
    }

    if let (Some(ev_type), Some(ev_index)) = (event_type, event_index) {
        if ev_type == "content_block_stop" {
            if let Some(kind) = state.block_kind_by_index.get(&ev_index) {
                if kind == "tool_use" {
                    if let Some(buffered) = state.tool_partial_json_by_index.get(&ev_index) {
                        let (normalized_json, normalize_reason) =
                            normalize_input_json_for_stream(buffered);

                        let mut shim_delta_payload = Map::new();
                        shim_delta_payload.insert(
                            "type".to_string(),
                            Value::String("content_block_delta".to_string()),
                        );
                        shim_delta_payload.insert(
                            "index".to_string(),
                            Value::Number(serde_json::Number::from(ev_index)),
                        );

                        let mut delta_inner = Map::new();
                        delta_inner.insert(
                            "type".to_string(),
                            Value::String("input_json_delta".to_string()),
                        );
                        delta_inner.insert(
                            "partial_json".to_string(),
                            Value::String(normalized_json.clone()),
                        );
                        shim_delta_payload.insert("delta".to_string(), Value::Object(delta_inner));

                        println!("[gateway sse shim] emitted_normalized_tool_input_delta=1 index={} reason={} normalized_len={}",
                            ev_index, normalize_reason, normalized_json.len());

                        output_data.push(serde_json::to_string(&shim_delta_payload).unwrap());
                        state.tool_partial_json_by_index.remove(&ev_index);
                    }
                }
            }
        }
    }

    output_data.push(payload);
    (event_name, output_data)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn frame(lines: &[&str]) -> Vec<String> {
        lines.iter().map(|s| s.to_string()).collect()
    }

    // --- 基础帧解析 ---

    #[test]
    fn empty_frame_returns_nothing() {
        let mut state = SseState::new();
        let (name, data) = process_sse_frame(&[], &mut state);
        assert!(name.is_none());
        assert!(data.is_empty());
    }

    #[test]
    fn event_only_frame_stores_pending_name() {
        let mut state = SseState::new();
        let f = frame(&["event: message_start"]);
        let (name, data) = process_sse_frame(&f, &mut state);
        assert!(name.is_none());
        assert!(data.is_empty());
        assert_eq!(state.pending_event_name, Some("message_start".into()));
    }

    #[test]
    fn data_frame_with_pending_event() {
        let mut state = SseState::new();
        state.pending_event_name = Some("content_block_start".into());
        let payload =
            json!({"type": "content_block_start", "index": 0, "content_block": {"type": "text"}});
        let f = frame(&[&format!("data: {}", payload)]);
        let (name, data) = process_sse_frame(&f, &mut state);
        assert_eq!(name, Some("content_block_start".into()));
        assert_eq!(data.len(), 1);
        assert!(state.pending_event_name.is_none());
    }

    #[test]
    fn done_frame() {
        let mut state = SseState::new();
        let f = frame(&["data: [DONE]"]);
        let (name, data) = process_sse_frame(&f, &mut state);
        assert_eq!(data, vec!["[DONE]"]);
    }

    #[test]
    fn malformed_json_emits_error_event() {
        let mut state = SseState::new();
        let f = frame(&["data: {not valid json"]);
        let (name, data) = process_sse_frame(&f, &mut state);
        assert_eq!(name, Some("error".into()));
        assert_eq!(data.len(), 1);
        let parsed: Value = serde_json::from_str(&data[0]).unwrap();
        assert_eq!(parsed["type"], "error");
    }

    // --- 合成 content_block_start ---

    #[test]
    fn delta_without_prior_start_synthesizes_block() {
        let mut state = SseState::new();
        let payload = json!({
            "type": "content_block_delta",
            "index": 0,
            "delta": {"type": "text_delta", "text": "hello"}
        });
        let f = frame(&[&format!("data: {}", payload)]);
        let (_name, data) = process_sse_frame(&f, &mut state);
        // 应该产生合成 start + 原始 delta
        assert!(
            data.len() >= 2,
            "expected synthetic start + delta, got {}",
            data.len()
        );
        let start: Value = serde_json::from_str(&data[0]).unwrap();
        assert_eq!(start["type"], "content_block_start");
        assert_eq!(start["index"], 0);
        assert!(state.seen_block_starts.contains(&0));
    }

    #[test]
    fn stop_without_prior_start_synthesizes_block() {
        let mut state = SseState::new();
        let payload = json!({
            "type": "content_block_stop",
            "index": 1
        });
        let f = frame(&[&format!("data: {}", payload)]);
        let (_name, data) = process_sse_frame(&f, &mut state);
        assert!(data.len() >= 2);
        let start: Value = serde_json::from_str(&data[0]).unwrap();
        assert_eq!(start["type"], "content_block_start");
        assert_eq!(start["index"], 1);
    }

    #[test]
    fn delta_with_prior_start_no_synthetic() {
        let mut state = SseState::new();
        state.seen_block_starts.insert(0);
        state.block_kind_by_index.insert(0, "text".into());
        let payload = json!({
            "type": "content_block_delta",
            "index": 0,
            "delta": {"type": "text_delta", "text": "hi"}
        });
        let f = frame(&[&format!("data: {}", payload)]);
        let (_name, data) = process_sse_frame(&f, &mut state);
        // 只有原始 payload，无合成 start
        assert_eq!(data.len(), 1);
    }

    // --- Tool input 缓冲 ---

    #[test]
    fn tool_use_delta_is_buffered() {
        let mut state = SseState::new();
        state.seen_block_starts.insert(0);
        state.block_kind_by_index.insert(0, "tool_use".into());

        let payload = json!({
            "type": "content_block_delta",
            "index": 0,
            "delta": {"type": "input_json_delta", "partial_json": "{\"q\":"}
        });
        let f = frame(&[&format!("data: {}", payload)]);
        let (_name, data) = process_sse_frame(&f, &mut state);
        // 被缓冲，不产生输出
        assert!(
            data.is_empty(),
            "tool delta should be buffered, got {:?}",
            data
        );
        assert_eq!(
            state.tool_partial_json_by_index.get(&0),
            Some(&"{\"q\":".to_string())
        );
    }

    #[test]
    fn tool_use_stop_emits_normalized_input() {
        let mut state = SseState::new();
        state.seen_block_starts.insert(0);
        state.block_kind_by_index.insert(0, "tool_use".into());
        state
            .tool_partial_json_by_index
            .insert(0, "{\"query\":\"test\"}".into());

        let payload = json!({
            "type": "content_block_stop",
            "index": 0
        });
        let f = frame(&[&format!("data: {}", payload)]);
        let (_name, data) = process_sse_frame(&f, &mut state);
        // 应产生 normalized delta + stop
        assert!(data.len() >= 2, "expected delta + stop, got {}", data.len());
        let delta: Value = serde_json::from_str(&data[0]).unwrap();
        assert_eq!(delta["type"], "content_block_delta");
        assert_eq!(delta["delta"]["type"], "input_json_delta");
        // buffer 应被清理
        assert!(state.tool_partial_json_by_index.get(&0).is_none());
    }

    #[test]
    fn tool_use_multiple_deltas_concatenated() {
        let mut state = SseState::new();
        state.seen_block_starts.insert(0);
        state.block_kind_by_index.insert(0, "tool_use".into());

        // 第一个 delta
        let p1 = json!({"type":"content_block_delta","index":0,"delta":{"type":"input_json_delta","partial_json":"{\"q\":"}});
        process_sse_frame(&frame(&[&format!("data: {}", p1)]), &mut state);

        // 第二个 delta
        let p2 = json!({"type":"content_block_delta","index":0,"delta":{"type":"input_json_delta","partial_json":"\"test\"}"}});
        process_sse_frame(&frame(&[&format!("data: {}", p2)]), &mut state);

        assert_eq!(
            state.tool_partial_json_by_index.get(&0),
            Some(&"{\"q\":\"test\"}".to_string())
        );
    }

    // --- Event name 传递 ---

    #[test]
    fn event_name_propagated_to_output() {
        let mut state = SseState::new();
        state.pending_event_name = Some("message_start".into());
        let payload = json!({"type": "message_start", "message": {}});
        let f = frame(&[&format!("data: {}", payload)]);
        let (name, _data) = process_sse_frame(&f, &mut state);
        assert_eq!(name, Some("message_start".into()));
    }
}
